### DISCUSSION TITLE
OpenID Connect example

### DISCUSSION THREAD

**Discussion by alxplusu** [maxcountryman/axum-login#253] [Ideas]
Hi!

first of all many thanks for the excellent library. I have been using it to set up authentication using OpenID Connect with the openidconnect-rs library. Would there by any use to add something like this as an additional example?

There is a lot of overlap with the OAuth2 example already there, but it might be helpful to have it as another example covering the OpenID Connect specifics since it is a fairly common use. If there is interest I would volunteer to clean up my code and make it something generic and minimal.

Best regards,
Alex Kastanas

**Comment by maxcountryman** [comment]
Glad to hear you've found it useful.

I think it could be helpful to add an additional example, so if you'd like to take that on I'd be happy to review.

**Reply by alxplusu** [reply]
Lovely! I'll follow the same structure as the OAuth2 one so they are compatible and people can compare and contrast. I'll make a pull request once I have a baseline implementation.

**Comment by alxplusu** [comment]
I have started to make the example here:

https://github.com/alxplusu/axum-login/tree/oidc-example/examples/oidc

I need to do some linting and cleanup before a PR as well as a sanity check that I am keeping all the needed information after the login process.

One question I have is whether in the case of OIDC we need to store the ID information in the database, since here we are primarily dealing with authentication. So my first thought was that it is enough to store the information in the session, but thought to check with you in advance to be sure. In case it is I can rework it to add back the sqlx.


Best regards,
Alex

**Reply by joshka** [reply]
This is usually a question that each app has to decide on their own whether they need. Some apps will need to store the users that login - if only to be able to map an external id onto an internal one easily. So it probably makes sense to make it part of the example (and make a note that it's an optional idea).

**Comment by joshka** [comment]
This example comes at a perfect time - I'd been searching for the best practice for implementing OIDC (something I'm pretty familiar with) with Axum (something I'm less familiar with). An example is a good idea, but there's a lot that goes into OIDC right, so perhaps extracting that logic makes sense. I saw a few different approaches e.g. the axum-oidc crate...

I'll definitely give this a read in the next little while, so thanks for writing this from a random axum user.

**Comment by alxplusu** [comment]
A good point about the SQL layer, makes sense. Thanks! And would appreciate a read, since OIDC is not my wheelhouse, come from a SAML background primarily.

I completely agree with your point and over the weekend I used also axum-oidc. It's really nice and easy, but it comes with the limitation that there is no authorization callback endpoint AFAIK. The identity provider is supposed to just redirect to next url directly. So in essence you need to be able to do wildcard redirect URLs or maintain a list of URLs (at least I think so). Which for what I have at hand is a no go.

But in the end I leaned on axum-login because I like the idea of separate packages taking care of the concept of logging in and then specific implementations for different approaches. 

I would like as a next step to make a package on top of axum-login with the OIDC logic,  but my Rust-fu is not there yet, still a newbie.

**Comment by JosiahParry** [comment]
Thanks @alxplusu this been a great help getting started! Though it seems like the example is a bit unfinished. For example, say we want to store the `profile` claims in the user. The `get_user()` endpoint doesn't return any of these claims. I'm a bit confused how to accomplish this since it seems like we don't have any way of storing the `id_token` or `access_token` in `axum-login`. 

Should the AuthSession backend be another tower-session and during the `authenticate()` method, the id_token and access_token be stored in the session?

**Comment by alxplusu** [comment]
Dear Josiah,

Thanks. I have had some evolution on this example, but life got in the way of getting the actual pull request done with the latest and greatest. 

I will take this as a reminder to make the PR and then we can take it from there. Just for the record I have added back the SQL backend part for keeping track of the user claim information, to be in line with other examples (and well, it makes sense, was unnecessary to remove it 😊 ).


Best, 
Alex

**Reply by JosiahParry** [reply]
Ah, okay that makes sense! If you push an updated commit please do let me know!

**Comment by GGORG0** [comment]
Hi, has there been any work done on this since? The example doesn't compile, but looks promising. I was just trying to implement OIDC here myself, but it's a bit difficult - for example how do you determine the session hash if the id token, access token and refresh token all get rotated. You'd have to verify the JWT signature of the id token or call userinfo - that's custom code and cannot be made into a hash that persists across token rotations.
